File: sdk/python/src/orcal/_transport.py
import time
import urllib.parse

import httpx

from . import errors

MAX_ATTEMPTS = 3
BACKOFF_SECONDS = 0.1
RETRYABLE_METHODS = frozenset({"GET", "HEAD"})


class Transport:
    def __init__(self, url, token, timeout=30.0, transport=None):
        self._base = url.rstrip("/")
        self._client = httpx.Client(timeout=timeout, transport=transport)
        self._token = token

    @staticmethod
    def escape(segment):
        return urllib.parse.quote(segment, safe="")

    def close(self):
        self._client.close()

    def _headers(self, extra=None):
        headers = {"Authorization": f"Bearer {self._token}"}
        if extra:
            headers.update(extra)
        return headers

    def _send(self, method, path, params, json, content, headers):
        return self._client.request(
            method,
            self._base + path,
            params=params,
            json=json,
            content=content,
            headers=self._headers(headers),
        )

    def request(self, method, path, *, params=None, json=None, content=None, headers=None):
        retryable = method.upper() in RETRYABLE_METHODS
        attempt = 0
        while True:
            attempt += 1
            try:
                response = self._send(method, path, params, json, content, headers)
            except httpx.TransportError:
                if not retryable or attempt >= MAX_ATTEMPTS:
                    raise
                time.sleep(BACKOFF_SECONDS * attempt)
                continue
            if response.status_code < 400:
                return response
            if response.status_code == 503 and retryable and attempt < MAX_ATTEMPTS:
                time.sleep(BACKOFF_SECONDS * attempt)
                continue
            raise errors.from_response(response.status_code, response.content)

    def stream(self, method, path, *, params=None, headers=None):
        return self._client.stream(
            method,
            self._base + path,
            params=params,
            headers=self._headers(headers),
        )

File: sdk/typescript/src/transport.ts
import { errorFromResponse } from "./errors.js";

export interface TransportOptions {
  url: string;
  token: string;
  timeoutMs?: number;
  fetch?: typeof fetch;
  backoffMs?: number;
}

export interface RequestOptions {
  params?: Record<string, string | number | undefined>;
  body?: unknown;
  raw?: BodyInit;
  headers?: Record<string, string>;
}

const MAX_ATTEMPTS = 3;
const RETRYABLE_METHODS = new Set(["GET", "HEAD"]);

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export class Transport {
  private readonly base: string;
  private readonly token: string;
  private readonly timeoutMs: number;
  private readonly backoffMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: TransportOptions) {
    this.base = options.url.replace(/\/+$/, "");
    this.token = options.token;
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.backoffMs = options.backoffMs ?? 100;
    this.fetchImpl = options.fetch ?? globalThis.fetch;
  }

  escape(segment: string): string {
    return encodeURIComponent(segment);
  }

  private url(path: string, params?: RequestOptions["params"]): string {
    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(params ?? {})) {
      if (value !== undefined) query.set(key, String(value));
    }
    const suffix = query.toString();
    return this.base + path + (suffix ? `?${suffix}` : "");
  }

  private init(method: string, options: RequestOptions): RequestInit {
    const headers: Record<string, string> = { authorization: `Bearer ${this.token}`, ...(options.headers ?? {}) };
    let body: BodyInit | undefined;
    if (options.raw !== undefined) {
      body = options.raw;
    } else if (options.body !== undefined) {
      body = JSON.stringify(options.body);
      headers["content-type"] = "application/json";
    }
    return { method, headers, body, signal: AbortSignal.timeout(this.timeoutMs) };
  }

  async request(method: string, path: string, options: RequestOptions = {}): Promise<Response> {
    const retryable = RETRYABLE_METHODS.has(method.toUpperCase());
    for (let attempt = 1; ; attempt += 1) {
      let response: Response;
      try {
        response = await this.fetchImpl(this.url(path, options.params), this.init(method, options));
      } catch (cause) {
        if (!retryable || attempt >= MAX_ATTEMPTS) throw cause;
        await sleep(this.backoffMs * attempt);
        continue;
      }
      if (response.status < 400) return response;
      if (response.status === 503 && retryable && attempt < MAX_ATTEMPTS) {
        await sleep(this.backoffMs * attempt);
        continue;
      }
      throw errorFromResponse(response.status, await response.text());
    }
  }

  async stream(path: string, params?: RequestOptions["params"]): Promise<Response> {
    const response = await this.fetchImpl(this.url(path, params), {
      method: "GET",
      headers: { authorization: `Bearer ${this.token}` },
    });
    if (response.status >= 400) throw errorFromResponse(response.status, await response.text());
    return response;
  }
}
